### PR TITLE
Create a helper hook for field validation

### LINKED ISSUE #37004: Create a helper hook for field validation
Blocked by https://github.com/woocommerce/woocommerce/issues/37007, https://github.com/woocommerce/woocommerce/issues/37005

Validation in blocks currently works by locking the post.  Locks are connected to a specific ID, so many locks can be added to prevent a post from being saved.

https://github.com/WordPress/gutenberg/blob/8a0eedeb85ba019f7179070f63447bb120e60308/docs/reference-guides/data/data-core-editor.md?plain=1#L1112

To make this easier, we should add some helper hooks around the locking for validation use cases.  The hook might look something like this:

```
const useValidation = ( id, validate ) => {
    const { lockPostSaving } = useDispatch( 'core/editor' );

    if ( ! validate( entity ) ) {
        lockPostSaving( id );
    } else {
        unlockPostSaving( id );
    }
}
```

And then in our block, we could pass in our own validation logic and ID:

```
const validateTitle = ( product ) => {
    if ( product.title.length < 2 ) {
        return false;
    }

    return true;
}

useValidation( 'product/title', validateTitle );
```

We should carefully consider the flexibility of this solution and make sure that it works well for our use case.  Some things to consider:

* Does passing the entity work for all use cases?
* Does this surface locks in a meaningful way?  Is there a chance the editor remains accidentally locked?
* Does this tie in well with error handling and render?  Will we need another util for that?

#### Acceptance criteria

* [ ] A validation hook that is reusable across all fields
* [ ] Post saving should be locked when invalid
* [ ] Should support async validation as well as sync

### PR DESCRIPTION
**Pull request by mdperez86**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/37004

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Pull branch `add/37004`.
2. Run `pnpm run test --filter=@woocommerce/product-editor`.
3. All tests should run successfully.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/product-editor/src/hooks/use-validation.ts
Comment (joshuatf) [inline packages/js/product-editor/src/hooks/use-validation.ts]: Is it possible to limit this to purely `boolean` or does `undefined` add value here?

File: packages/js/product-editor/src/hooks/use-validation.ts
Comment (joshuatf) [inline packages/js/product-editor/src/hooks/use-validation.ts]: Love these examples!  It might be worthwhile to move these to a MD file for discoverability and easier parsing once the docs project gets underway.

File: packages/js/product-editor/jest.config.json
Comment (joshuatf) [inline packages/js/product-editor/jest.config.json:1]: Nice!  Just a heads up that we'll probably have conflicts (or hopefully just need a rebase) if [this](https://github.com/woocommerce/woocommerce/pull/37225) is merged first, but we'll probably merge your PR first anyway.

File: packages/js/data/src/user/use-user-preferences.ts
Comment (joshuatf) [inline packages/js/data/src/user/use-user-preferences.ts]: I think removing these is causing issues in the CI workflow:
Screen Shot 2023-03-14 at 5 31 22 PM

**Review by joshuatf — COMMENTED** [review]
Nice work!  Left a few comments/questions.

Also do you think we should include validation for the product name directly in this PR so we have a manual test case or did you prefer to hook that up later?

File: packages/js/product-editor/src/hooks/use-validation.ts
Comment (mdperez86) [reply packages/js/product-editor/src/hooks/use-validation.ts]: I did that intentionally because we don't want to validate a field if it's not dirty or touched, so I left the `undefined` there to cover those scenarios. If we remove it then we will need other custom hooks to capture dirty/touched states.

File: packages/js/product-editor/src/hooks/use-validation.ts
Comment (mdperez86) [reply packages/js/product-editor/src/hooks/use-validation.ts]: Thanks! Do you want me to move this examples to a MD file in this PR?
If yes:
Where the MD file should be located?
What should be the name?

File: packages/js/product-editor/jest.config.json
Comment (mdperez86) [reply packages/js/product-editor/jest.config.json:1]: Thanks, it as totally expected!

File: packages/js/data/src/user/use-user-preferences.ts
Comment (mdperez86) [reply packages/js/data/src/user/use-user-preferences.ts]: The only reason I removed these it was because in my local env was not working with them and started to work after I removed them. I had to take another look at why locally worked but not remotely.
Thanks

Comment (mdperez86) [reply packages/js/data/src/user/use-user-preferences.ts]: done -> [61c366d](https://github.com/woocommerce/woocommerce/pull/37196/commits/61c366da2f837278bee0a5e808c88e725b2d11ba)

File: packages/js/product-editor/src/hooks/use-validation.ts
Comment (joshuatf) [reply packages/js/product-editor/src/hooks/use-validation.ts]: I think creating a folder for `use-validation` and then adding this to a `README.md` file may be a good pattern.  Gutenberg sometimes does this for hooks - https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-instance-id/README.md

**Review by joshuatf — DISMISSED** [review]
Awesome work, @mdperez86!  Pre-approving pending the move of the documentation to a markdown file and passing tests, but this looks great.

File: packages/js/product-editor/src/hooks/use-validation/README.md
Comment (joshuatf) [inline packages/js/product-editor/src/hooks/use-validation/README.md:9]: I think at the DM we may have mentioned avoiding using TS for readmes and examples and documentation to reduce the barrier to entry for 3PD, but we can update this in a follow-up if that's how we want to handle this.  No need to hold up this PR.

**Review by joshuatf — APPROVED** [review]
Added one comment I missed before, but this is more long-term discussion on how we want to handle docs/examples.

This PR LGTM! 🚢

**Pull request merged**
